### ISSUE TITLE
Could not reproduce your experiment

### ISSUE THREAD

**Issue by RunZH** [AlexeyAB/darknet#52]
Hi, I was trying to reproduce your YOLO v2 experiment. But I found the IOU and Recall are much lower than yours(yours are 74.61% and 90.12%, Mine is 56.44% and 69.23%). I don't know how to use vocdevkit so I can't test mAP to compare with yours. Could you please show more details on your training setting?
what's more, how did you calculate the mAP for testing set?

My setting is:
darknet19 as pretrained weights, Pascal VOC Data and I didn't use

**Comment by AlexeyAB** [comment]
Hi, @RunZH 

* Do you use Windows or Linux darknet?
* Did you set `random=1`: https://github.com/AlexeyAB/darknet/blob/master/build/darknet/x64/yolo-voc.cfg#L244
* Did you use cuDNN?
* Did you do all steps? https://github.com/AlexeyAB/darknet#how-to-train-pascal-voc-data
    Mainly: 5. Run command: `type 2007_train.txt 2007_val.txt 2012_*.txt > train.txt`

**Comment by RunZH** [comment]
I use the Windows darknet from you.
I follow all steps in https://github.com/AlexeyAB/darknet#how-to-train-pascal-voc-data 
Add 2007 train, val and all 2012 to training set.  After I calculate mAP. I found it was only 0.4573. 
After I use random=1 it was even worse, 0.3937.
(as for the download weights i got mAP of  0.7641)

I used the version of commits on Dec 13, 2016. The cfg use : 
learning_rate=0.0001
max_batches = 45000
policy=steps
steps=100,25000,35000
scales=10,.1,.1
and I found you updated the cfg in last 4 days. Should that be the problem for me? Is that why I can't get high mAP?

**Comment by AlexeyAB** [comment]
> and I found you updated the cfg in last 4 days. Should that be the problem for me? Is that why I can't get high mAP?

May be. Basically, this improved the accuracy for COCO.
Updated cfg-files in this fork and weights-files by link: https://pjreddie.com/darknet/yolo/

> I use the Windows darknet from you.
> I used the version of commits on Dec 13, 2016. 

What version of commits did you use? Since Dec 13, 2016 in this fork there were fixed some bugs, which strongly affect the accuracy.

**Comment by RunZH** [comment]
I used the version Committed on Mar 9, 2017.

I want to reproduce the experiment to get the result of the paper which is mAP = 76.8. I directly use the setting in the yolo_voc.cfg with: 
batch=64
subdivisions=8
learning_rate=0.0001
max_batches = 45000
policy=steps
steps=100,25000,35000
scales=10,.1,.1
random = 1
(now it has been put at https://github.com/pjreddie/darknet/blob/master/cfg/yolo-voc.2.0.cfg by pjreddie
and I think this setting is different from the paper)

I have 16552 training images (contains train2012, val2012, train2007 val2007) and 4953 testing images
use darknet19 as pre-trained weights.

I wrote a function to calculate mAP using the data produced by cmd: 
darknet.exe detector valid data/voc.data yolo-voc.cfg  yolo-voc_45000.weights -i 1
If the training code works well. I think it should at least give me mAP over 70. But the mAP is around 45.
I also observed the detection output on the test set. A lot of objects were not detected by the model trained by me. 
I wonder if there is anything missed in this training set up which prevents me from getting the published result?

**Comment by AlexeyAB** [comment]
@RunZH 

There was bug of random generator on Windows:

At Commits on Mar 14, 2017: https://github.com/AlexeyAB/darknet/commits/master

>  @AlexeyAB
> Fixed training with rand_s()

This greatly influenced the quality of training: https://github.com/AlexeyAB/darknet/issues/19#issuecomment-286476902

----

You should update source code from this fork, use old cfg-file: https://github.com/pjreddie/darknet/blob/master/cfg/yolo-voc.2.0.cfg

Set `random=1` and train again using `darknet19_448.conv.23`.

**Comment by RunZH** [comment]
@AlexeyAB 
Thanks! The mAP has improved after I use the new commit. I will update the result after the training is done.

By the way, is there any way to speed up training? It takes me about 1.5 hours to traing 1000 iters on one NVIDIA GTX 1080. And it is unstable on my PC, it makes my PC reboot frequently when I am training.

**Comment by AlexeyAB** [comment]
@RunZH To speed up training 

- you can use cuDNN - point 5: https://github.com/AlexeyAB/darknet#how-to-compile
- and 4 x GTX 1080 :) https://github.com/AlexeyAB/darknet#how-to-train-with-multi-gpu

Yolo should not reboot PC, so I don't know why this is happen.

**Comment by RunZH** [comment]
@AlexeyAB 
I have finished training and the mAP is 75.6 which is lower than the downloaded weights (76.4) on Marth 10.
Is it possible?

Thanks!

**Comment by AlexeyAB** [comment]
@RunZH  I think yes, it possible. Training uses random values so the result each time may differ slightly.
It is also possible that Redmon changed some more values in the cfg file.
